Lib/fontbakery/profiles/layout.py

```diff
@@ -0,0 +1,113 @@
+from fontbakery.callable import check
+from fontbakery.status import PASS, FAIL
+from fontbakery.message import Message
+from fontbakery.fonts_profile import profile_factory # NOQA pylint: disable=unused-import
+
+
+from opentypespec.tags import FEATURE_TAGS, SCRIPT_TAGS, LANGUAGE_TAGS
+
+
+def feature_tags(ttFont):
+    in_this_font = set()
+    for table in ["GSUB", "GPOS"]:
+        if ttFont.get(table) and ttFont[table].table.FeatureList:
+            for fr in ttFont[table].table.FeatureList.FeatureRecord:
+                in_this_font.add(fr.FeatureTag)
+    return in_this_font
+
+
+DEPRECATED_TAGS = ["hngl", "opbd", "size"]
+
+
+@check(
+    id = "com.google.fonts/check/layout_valid_feature_tags",
+    rationale = """
+        Incorrect tags can be indications of typos, leftover debugging code or questionable approaches, or user error in the font editor. Such typos can cause features and language support to fail to work as intended.
+    """,
+    misc_metadata = {
+        "request": ["https://github.com/googlefonts/fontbakery/issues/3355"],
+        "severity": 8,
+    },
+)
+def com_google_fonts_check_layout_valid_feature_tags(ttFont):
+    """Does the font have any invalid feature tags?"""
+    bad_tags = set()
+    for tag in feature_tags(ttFont):
+        if tag not in FEATURE_TAGS.keys():
+            bad_tags.add(tag)
+    if bad_tags:
+        yield FAIL, \
+              Message("bad-feature-tags",
+                      "The following invalid feature tags were found in the font: "
+                      + ", ".join(bad_tags))
```

Comment:
This shouldn’t be a fail IMO. We are about to ship a font that uses a private HarfBuzz feature tag (`HARF` and `BUZZ` tags) and it will fail this test even though it is a perfectly valid font. Feature tag registry is not part of the OpenType spec and fonts/layout engines are free to use other tags (Chrome/Firefox/TeX/LibreOffice will happily allow one to activate any feature tag). The test should be split into two, a fail for really invalid feature tags (e.g. empty ones) and a warning for unregistered feature tags.

Reply:
It may be a good idea to downgrade *some* of these fails to warns, but the general sentiment that fontbakery should not FAIL a "perfectly valid font" is not true. We are not testing compliance with the OT Spec, but a rather opinionated view of what an OT font should contain. (For example, we fail if 0x00A0 has glyph name "nbsp", or if Indic fonts don't have a rupee symbol.)

My worry is that once you start trying to separate out what's an "other tag" and what's definitely a mistake, you make a very hard job for yourself and end up missing some mistakes which should definitely be fails.

Reply:
May be the wrong choice, but here the private features are not superfluous or extra, but a required part of the font without it the font would not function correctly, despite not being a registered feature.

Reply:
I trust that you have good reasons for making a font which only works correctly on one shaping engine and not on any others, but I don't think you should expect that such a font should pass a fontbakery check. :-)

Maybe disable that check in your QA run.

Reply:
It is the other way around. HarfBuzz has a bug/limitation other layout engines don’t have, and by using the private feature we workaround this and make the font behave consistently. It would have been OK if I was voluntarily using fontbakery (I don’t, never had a good experience using it), but it is forced upon me by and it will be on me to explain every time why my fonts are correct and fontbakery is being pedantic for no good reason.

An unregistered feature is at worst some extra unused bytes, it does not affect functionality in any way. Flagging this as a FAIL with potential false +ve is way too much.

Reply:
OK. @davelab6, @felipesanches, opinions?

Reply:
if it is not a strictly OpenType spec check, then it may be better to be placed on the `universal` profile. Also, if it is too much of an opinionated check, then it might be better to be placed it on a vendor-specific one (in this case, the `googlefonts` profile).

Reply:
> "never had a good experience using it"

@khaledhosny, I want to change this. Would you like to schedule a video call to provide feedback on the tool? I would be glad to learn in more detail how can I improve the user experience.

Reply:
> "I don't think you should expect that such a font should pass a fontbakery check. :-)"

@simoncozens, this is a too broad statement. I think that your comment would be more accurate (to reflect the Font Bakery project policies) if you restricted it to a more specific case such as: "you shouldn't expect that such a font would pass a `googlefonts` profile check"

Reply:
> "Maybe disable that check in your QA run."

Yeah, this is one way of adding exceptions to fontbakery checking. You can set-up the QA step in your workflow to ignore specific checks by using the `-x / --exclude-checkid` flag (which can be used multiple times if you need to exclude multiple checks).

But it would be better to address @khaledhosny 's concerns so that excluding checks would be kept as a last resort instead of as a wide-spread practice. We want fontbakery to be useful "as is" to the broadest possible set of users. That's one of the reasons why we group checks in profiles.

Reply:
> @khaledhosny, I want to change this. Would you like to schedule a video call to provide feedback on the tool? I would be glad to learn in more detail how can I improve the user experience.

FontBakery is just too opinionated for me, and I happen to almost always build fonts in “unconventional” ways. What I’d be interested in would be something that reports actual bugs not best practices and finding actual bugs that would otherwise go unnoticed during normal use of the font is rather hard and probably need lots of font-specific knowledge so a general tool like FB wouldn’t help me much. I’ve to use FB when I submit/update fonts on Google Fonts, and that is usually the most annoying because most of the failing checks are rather arbitrary things with dubious rationales (to me at least) that GF insist on.

Reply:
@khaledhosny, please open individual issues to discuss the rationales that you consider dubious and we'll try to accommodate.